File: javascript/MaximumSubarray.js
// Time: O(n), Space: O(1)
var maxSubArray = function(nums) {
    let maxSoFar = nums[0];
    let current = nums[0];
    for (let i = 1; i < nums.length; i++) {
        current = Math.max(nums[i], current + nums[i]);
        maxSoFar = Math.max(maxSoFar, current);
    }
    return maxSoFar;
};

File: c++/Maximum Subarray.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxCurrent = nums[0], maxGlobal = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            maxCurrent = max(nums[i], maxCurrent + nums[i]);
            maxGlobal = max(maxGlobal, maxCurrent);
        }
        return maxGlobal;
    }
};
